paho/client.go

```diff
@@ -127,154 +138,167 @@ func NewClient(conf ClientConfig) *Client {
-	go func() {
-		defer c.workers.Done()
-		c.writer()
-	}()
+		go c.writer()
+		go c.reader()
 
-	debug.Println("starting reader")
-	c.workers.Add(1)
-	go func() {
-		defer c.workers.Done()
-		c.reader()
-	}()
+		connCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
+		defer cf()
+		c.caCtx = &caContext{connCtx, make(chan *packets.Connack, 1)}
+		defer func() {
+			c.caCtx = nil
+		}()
 
-	connCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
-	defer cf()
-	c.caCtx = &caContext{connCtx, make(chan *packets.Connack, 1)}
-	defer func() {
-		c.caCtx = nil
-	}()
-
-	ccp := cp.Packet()
+		ccp := cp.Packet()
+		ccp.ProtocolName = "MQTT"
+		ccp.ProtocolVersion = 5
 
-	ccp.ProtocolName = "MQTT"
-	ccp.ProtocolVersion = 5
+		debug.Println("sending CONNECT")
+		if c.cerr = c.write(ctx, ccp); c.cerr != nil {
+			return
+		}
 
-	debug.Println("sending CONNECT")
-	if err := c.write(ctx, ccp); err != nil {
-		cleanup()
-		return nil, err
-	}
+		debug.Println("waiting for CONNACK")
+		var cap *packets.Connack
+		select {
+		case <-connCtx.Done():
+			debug.Println("waiting for CONNACK canceled")
+			c.cerr = connCtx.Err()
+			return
 
-	debug.Println("waiting for CONNACK")
-	var cap *packets.Connack
-	select {
-	case <-connCtx.Done():
-		debug.Println("waiting for CONNACK canceled")
-		cleanup()
-		return nil, connCtx.Err()
-	case cap = <-c.caCtx.Return:
-	}
+		case cap = <-c.caCtx.Return:
+			debug.Println("received CONNACK")
+		}
 
-	ca := ConnackFromPacketConnack(cap)
+		ca := ConnackFromPacketConnack(cap)
 
-	if ca.ReasonCode >= 0x80 {
-		var reason string
-		debug.Println("received an error code in Connack:", ca.ReasonCode)
-		if ca.Properties != nil {
-			reason = ca.Properties.ReasonString
+		if ca.ReasonCode >= 0x80 {
+			var reason string
+			debug.Println("received an error code in Connack:", ca.ReasonCode)
+			if ca.Properties != nil {
+				reason = ca.Properties.ReasonString
+			}
+			c.cerr = fmt.Errorf("failed to connect to server: %s", reason)
+			return
 		}
-		cleanup()
-		return ca, fmt.Errorf("failed to connect to server: %s", reason)
-	}
 
-	if ca.Properties != nil {
-		if ca.Properties.ServerKeepAlive != nil {
-			keepalive = *ca.Properties.ServerKeepAlive
-		}
-		if ca.Properties.AssignedClientID != "" {
-			c.ClientID = ca.Properties.AssignedClientID
-		}
-		if ca.Properties.ReceiveMaximum != nil {
-			c.serverProps.ReceiveMaximum = *ca.Properties.ReceiveMaximum
-		}
-		if ca.Properties.MaximumQoS != nil {
-			c.serverProps.MaximumQoS = *ca.Properties.MaximumQoS
-		}
-		if ca.Properties.MaximumPacketSize != nil {
-			c.serverProps.MaximumPacketSize = *ca.Properties.MaximumPacketSize
-		}
-		if ca.Properties.TopicAliasMaximum != nil {
-			c.serverProps.TopicAliasMaximum = *ca.Properties.TopicAliasMaximum
+		if ca.Properties != nil {
+			if ca.Properties.ServerKeepAlive != nil {
+				keepalive = *ca.Properties.ServerKeepAlive
+			}
+			//if ca.Properties.AssignedClientID != "" {
+			//	c.ClientID = ca.Properties.AssignedClientID
+			//}
```

Comment:
How come you've commented this our, seems to be part of the `CONNACK` [specs](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901087). Are you planning to fix it in a coming PR?

Reply:
This client id is not needed anywhere.